libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -0,0 +1,580 @@
+	for (auto const& [type, funNames]: types)
+	{
+		for (string const& name: funNames)
+			m_functionTypes[name] = index;
+		result += toBytes(ValueType::Function);
+		result += lebEncode(type.first.size()) + type.first;
+		result += lebEncode(type.second.size()) + type.second;
+
+		index++;
+	}
+
+	return makeSection(Section::TYPE, lebEncode(index) + std::move(result));
+}
+
+bytes BinaryTransform::importSection(
+	vector<FunctionImport> const& _imports
+)
+{
+	bytes result = lebEncode(_imports.size());
+	for (FunctionImport const& import: _imports)
+	{
+		uint8_t importKind = 0; // function
+		result +=
+			encodeName(import.module) +
+			encodeName(import.externalName) +
+			toBytes(importKind) +
+			lebEncode(m_functionTypes[import.internalName]);
+	}
+	return makeSection(Section::IMPORT, std::move(result));
+}
+
+bytes BinaryTransform::functionSection(vector<FunctionDefinition> const& _functions)
+{
+	bytes result = lebEncode(_functions.size());
+	for (auto const& fun: _functions)
+		result += lebEncode(m_functionTypes.at(fun.name));
+	return makeSection(Section::FUNCTION, std::move(result));
+}
+
+bytes BinaryTransform::memorySection()
+{
+	bytes result = lebEncode(1);
+	result.push_back(0); // flags
+	result.push_back(1); // initial
+	return makeSection(Section::MEMORY, std::move(result));
+}
+
+bytes BinaryTransform::globalSection()
+{
+	bytes result = lebEncode(m_globals.size());
+	for (size_t i = 0; i < m_globals.size(); ++i)
+		result +=
+			// mutable i64
+			bytes{uint8_t(ValueType::I64), 1} +
+			toBytes(Opcode::I64Const) +
+			lebEncodeSigned(0) +
+			toBytes(Opcode::End);
+
+	return makeSection(Section::GLOBAL, std::move(result));
+}
+
+bytes BinaryTransform::exportSection()
+{
+	bytes result = lebEncode(2);
+	result += encodeName("memory") + toBytes(Export::Memory) + lebEncode(0);
+	result += encodeName("main") + toBytes(Export::Function) + lebEncode(m_functions.at("main"));
+	return makeSection(Section::EXPORT, std::move(result));
+}
+
+bytes BinaryTransform::customSection(string const& _name, bytes _data)
+{
+	bytes result = encodeName(_name) + std::move(_data);
+	return makeSection(Section::CUSTOM, std::move(result));
+}
+
+bytes BinaryTransform::codeSection(vector<wasm::FunctionDefinition> const& _functions)
+{
+	bytes result = lebEncode(_functions.size());
+	for (FunctionDefinition const& fun: _functions)
+		result += (*this)(fun);
+	return makeSection(Section::CODE, std::move(result));
+}
+
+bytes BinaryTransform::visit(vector<Expression> const& _expressions)
+{
+	bytes result;
+	for (auto const& expr: _expressions)
+		result += boost::apply_visitor(*this, expr);
+	return result;
+}
+
+bytes BinaryTransform::visitReversed(vector<Expression> const& _expressions)
+{
+	bytes result;
+	for (auto const& expr: _expressions | boost::adaptors::reversed)
+		result += boost::apply_visitor(*this, expr);
+	return result;
+}
+
+bytes BinaryTransform::encodeName(std::string const& _name)
+{
+	// Should be utf-8 encoded.
```

Comment:
Don't want to nitpick but we do have `libdevcore/UTF8.h` with `validateUTF8` which could be used here :)

Reply:
That's fine, just amend the comment to say "UTF-8 is allowed here by the Wasm spec, but we want to restrict ourselves to ASCII only".